code/game/objects/structures/noticeboard.dm

```diff
@@ -55,54 +58,74 @@
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "NoticeBoard", name)
+		ui.open()
+
+/obj/structure/noticeboard/ui_data(mob/user)
+	var/list/data = list()
+	data["allowed"] = allowed(user)
+	data["items"] = list()
+	for(var/obj/item/content in contents)
+		var/list/content_data = list(
+			name = content.name,
+			ref = REF(content)
+		)
+		data["items"] += list(content_data)
+	return data
+
+/obj/structure/noticeboard/ui_act(action, params)
 	. = ..()
-	var/auth = allowed(user)
-	var/dat = "<B>[name]</B><BR>"
-	for(var/obj/item/P in src)
-		if(istype(P, /obj/item/paper))
-			dat += "<A href='?src=[REF(src)];read=[REF(P)]'>[P.name]</A> [auth ? "<A href='?src=[REF(src)];write=[REF(P)]'>Write</A> <A href='?src=[REF(src)];remove=[REF(P)]'>Remove</A>" : ""]<BR>"
-		else
-			dat += "<A href='?src=[REF(src)];read=[REF(P)]'>[P.name]</A> [auth ? "<A href='?src=[REF(src)];remove=[REF(P)]'>Remove</A>" : ""]<BR>"
-	user << browse("<HEAD><TITLE>Notices</TITLE></HEAD>[dat]","window=noticeboard")
-	onclose(user, "noticeboard")
-
-/obj/structure/noticeboard/Topic(href, href_list)
-	..()
-	usr.set_machine(src)
-	if(href_list["remove"])
-		if(usr.stat != CONSCIOUS || HAS_TRAIT(usr, TRAIT_HANDS_BLOCKED)) //For when a player is handcuffed while they have the notice window open
-			return
-		var/obj/item/I = locate(href_list["remove"]) in contents
-		if(istype(I) && I.loc == src)
-			I.forceMove(usr.loc)
-			usr.put_in_hands(I)
-			notices--
-			icon_state = "nboard0[notices]"
+	if(.)
+		return
 
-	if(href_list["write"])
-		if(usr.stat != CONSCIOUS || HAS_TRAIT(usr, TRAIT_HANDS_BLOCKED)) //For when a player is handcuffed while they have the notice window open
-			return
-		var/obj/item/P = locate(href_list["write"]) in contents
-		if(istype(P) && P.loc == src)
-			var/obj/item/I = usr.is_holding_item_of_type(/obj/item/pen)
-			if(I)
-				add_fingerprint(usr)
-				P.attackby(I, usr)
-			else
-				to_chat(usr, "<span class='warning'>You'll need something to write with!</span>")
+	var/obj/item/item = locate(params["ref"]) in contents
+	if(!istype(item) || item.loc != src)
+		return
 
-	if(href_list["read"])
-		var/obj/item/I = locate(href_list["read"]) in contents
-		if(istype(I) && I.loc == src)
-			usr.examinate(I)
+	var/mob/user = usr
+
+	switch(action)
+		if("examine")
+			if(istype(item, /obj/item/paper))
+				item.ui_interact(user)
+			else
+				user.examinate(item)
+			return TRUE
+		if("remove")
+			if(!allowed(user))
+				return
+			remove_item(item, user)
+			return TRUE
+
+/**
+ * Removes an item from the notice board
+ *
+ * Arguments:
+ * * item - The item that is to be removed
+ * * user - The mob that is trying to get the item removed, if there is one
+ */
+/obj/structure/noticeboard/proc/remove_item(obj/item/item, mob/user)
+	if(istype(item))
+		if(user)
+			item.forceMove(user.loc)
+			user.put_in_hands(item)
+			to_chat(user, "<span class='notice'>You remove [item] from [src].</span>")
```

Comment:
```suggestion
			balloon_alert(user, "removed from board")
```

We are trying to avoid chat procs for things that are not important for archival.